File: src/augmentedquill/models/content_recovery.py
# Copyright (C) 2026 StableLlama
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

"""Transport models for explicit, revision-guarded content recovery."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class ContentRecoverySummary(BaseModel):
    """Validated metadata for one durable content recovery pair."""

    recovery_id: str
    status: Literal["prepared", "committed"]
    document_key: str
    filename: str
    before_revision: str
    after_revision: str
    created_at: str
    committed_at: str | None = None


class ContentRecoveryDetail(ContentRecoverySummary):
    """A recovery pair including exact before and after manuscript text."""

    before_content: str
    after_content: str


class ContentRecoveryListResponse(BaseModel):
    """Response for the project recovery listing."""

    records: list[ContentRecoverySummary]


class ContentRecoveryRestoreRequest(BaseModel):
    """Explicit restore target and the caller's current checked base."""

    target: Literal["before", "after"]
    expected_revision: str = Field(min_length=1)
    expected_document_key: str = Field(min_length=1)
    expected_filename: str | None = None


class ContentRecoveryRestoreResponse(BaseModel):
    """Canonical content returned after a successful restore."""

    ok: bool
    content: str
    revision: str
    filename: str
    document_key: str

File: src/frontend/services/apiClients/contentRecovery.ts
// Copyright (C) 2026 StableLlama
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

/** Purpose: Access explicit, revision-guarded durable content recovery. */
import { fetchJson, postJson, projectEndpoint } from './shared';

export interface ContentRecoverySummary {
  recovery_id: string;
  status: 'prepared' | 'committed';
  document_key: string;
  filename: string;
  before_revision: string;
  after_revision: string;
  created_at: string;
  committed_at?: string | null;
}

export interface ContentRecoveryDetail extends ContentRecoverySummary {
  before_content: string;
  after_content: string;
}

export interface ContentRecoveryRestoreOptions {
  target: 'before' | 'after';
  expected_revision: string;
  expected_document_key: string;
  expected_filename?: string;
}

export interface ContentRecoveryRestoreResponse {
  ok: boolean;
  content: string;
  revision: string;
  filename: string;
  document_key: string;
}

export interface ContentRecoveryApi {
  list: () => Promise<{ records: ContentRecoverySummary[] }>;
  get: (recoveryId: string) => Promise<ContentRecoveryDetail>;
  restore: (
    recoveryId: string,
    options: ContentRecoveryRestoreOptions
  ) => Promise<ContentRecoveryRestoreResponse>;
}

function recoveryEndpoint(projectName: string, recoveryId?: string): string {
  const suffix = recoveryId
    ? `/content-recovery/${encodeURIComponent(recoveryId)}`
    : '/content-recovery';
  return projectEndpoint(projectName, suffix);
}

export const createContentRecoveryApi = (projectName: string): ContentRecoveryApi => ({
  list: async (): Promise<{ records: ContentRecoverySummary[] }> =>
    fetchJson<{ records: ContentRecoverySummary[] }>(
      recoveryEndpoint(projectName),
      undefined,
      'Failed to list content recovery records'
    ),

  get: async (recoveryId: string): Promise<ContentRecoveryDetail> =>
    fetchJson<ContentRecoveryDetail>(
      recoveryEndpoint(projectName, recoveryId),
      undefined,
      'Failed to get content recovery record'
    ),

  restore: async (
    recoveryId: string,
    options: ContentRecoveryRestoreOptions
  ): Promise<ContentRecoveryRestoreResponse> =>
    postJson<ContentRecoveryRestoreResponse>(
      `${recoveryEndpoint(projectName, recoveryId)}/restore`,
      options,
      'Failed to restore content recovery record'
    ),
});

export const contentRecoveryApi = createContentRecoveryApi('');
